<a href="https://colab.research.google.com/github/bolouie/Google-Play-Android-App-Analysis/blob/main/Google_Play_Store_App_Analytics_(Complete).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

A comprehensive analysis of the Android app market by comparing thousands of apps in the Google Play store.

# About the Dataset of Google Play Store Apps & Reviews

**Data Source:** <br>
App and review data was scraped from the Google Play Store by Lavanya Gupta in 2018. Original files listed [here](
https://www.kaggle.com/lava18/google-play-store-apps).

# Import Statements

In [ ]:
import pandas as pd


# Notebook Presentation

In [ ]:
# Show numeric output in decimal format e.g., 2.15
pd.options.display.float_format = '{:,.2f}'.format

# Read the Dataset

In [ ]:
df_apps = pd.read_csv('apps.csv')

# Data Cleaning

**Challenge**: How many rows and columns does `df_apps` have? What are the column names? Look at a random sample of 5 different rows with [.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html).

In [ ]:
df_apps.shape

(10841, 12)

In [ ]:
df_apps.sample(20)

App            Category  \
9605                                       Crazy Doctor              FAMILY   
7256                     Calendar+ Schedule Planner App        PRODUCTIVITY   
5621                               HTC Sense Input - FA               TOOLS   
9636                     Zapya - File Transfer, Sharing               TOOLS   
10500                     ZEDGE™ Ringtones & Wallpapers     PERSONALIZATION   
3469                                Light X - Icon Pack     PERSONALIZATION   
527                                            quran-DZ              SOCIAL   
5282                                          CWT To Go    TRAVEL_AND_LOCAL   
1138                      Dulquer Salmaan HD Wallpapers     PERSONALIZATION   
3643   3DAnimeGirl DX DreamPortrait KAWAII Girl DressUp              FAMILY   
9878                                         Stack Jump                GAME   
10103                                    Kick the Buddy                GAME   
10133                    InstaBeauty -Makeup Selfie Cam         PHOTOGRAPHY   
393                                 EB Demo for Android              FAMILY   
6390                         Find&Save - Local Shopping            SHOPPING   
430                                           AQ Guards        PRODUCTIVITY   
8381                                          USA TODAY  NEWS_AND_MAGAZINES   
1524                                          BC Camera         PHOTOGRAPHY   
2590                                    Alerte au gogol              FAMILY   
219                                            DS-Admin              FAMILY   

       Rating  Reviews  Size_MBs     Installs  Type  Price Content_Rating  \
9605     4.00   267636     13.00   10,000,000  Free      0           Teen   
7256     4.00     8985      5.70    1,000,000  Free      0       Everyone   
5621     4.60     1420      1.60      100,000  Free      0       Everyone   
9636     4.60   387958      9.00   10,000,000  Free      0           Teen   
10500    4.60  6459626      6.15  100,000,000  Free      0           Teen   
3469     4.60      252     24.00       10,000  Paid  $0.99       Everyone   
527       NaN        0      6.20           10  Free      0           Teen   
5282     4.40    20313     57.00      100,000  Free      0       Everyone   
1138      NaN        1      3.90          100  Free      0       Everyone   
3643     4.20      449     56.00       10,000  Free      0           Teen   
9878     4.30   118253     40.00   10,000,000  Free      0           Teen   
10103    4.30  1004709     36.00   50,000,000  Free      0           Teen   
10133    4.30   654419      6.90   50,000,000  Free      0       Everyone   
393       NaN        0     13.00           10  Free      0           Teen   
6390     4.00     4602      6.20      500,000  Free      0       Everyone   
430       NaN        0      3.20           10  Free      0       Everyone   
8381     4.10    49259      6.30    5,000,000  Free      0   Everyone 10+   
1524     4.20        5      5.50          500  Free      0       Everyone   
2590      NaN       14      2.00        1,000  Free      0       Everyone   
219       NaN        1     21.00           10  Free      0       Everyone   

                 Genres       Last_Updated         Android_Ver  
9605       Role Playing      March 3, 2014          2.1 and up  
7256       Productivity   January 16, 2017        4.0.3 and up  
5621              Tools   November 9, 2015          5.0 and up  
9636              Tools      July 16, 2018        4.0.3 and up  
10500   Personalization      July 19, 2018  Varies with device  
3469    Personalization      June 13, 2018          4.1 and up  
527              Social      June 13, 2018          4.2 and up  
5282     Travel & Local     August 5, 2018          4.4 and up  
1138    Personalization       May 22, 2018          4.0 and up  
3643             Casual  February 16, 2018          4.0 and up  
9878             Arcade        May 3, 2018          4.1 and

### Drop Unused Columns

**Challenge**: Remove the columns called `Last_Updated` and `Android_Version` from the DataFrame. We will not use these columns. 

Provide a list of the columns names `['Last_Update, 'Android_Ver']` to the `.drop()` method. By setting axis=1, this specifies the columns to drop. 

In [ ]:
df_apps.drop(['Last_Updated', 'Android_Ver'], axis=1, inplace=True)
df_apps.head()

App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  
0        0      8.70        0  Paid  $13.99           Teen           Social  
1        0     33.00        0  Paid   $2.99       Everyone        Education  
2        0      5.50        0  Paid   $1.49       Everyone  Personalization  
3        0     19.00        0   NaN       0   Everyone 10+         Strategy  
4        0     10.00        0  Free       0       Everyone         Business

### Find and Remove NaN values in Ratings

**Challenge**: How may rows have a NaN (not-a-number) value in the Ratings column? Create DataFrame called `df_apps_clean` that does not include these rows. 

To find and remove the rows with the NaN values we can create a subset of the DataFrame based on where `.isna()` evaluates to True. 

In [ ]:
nan_rows = df_apps[df_apps.Rating.isna()]
print(nan_rows.shape)
nan_rows.head()

(1474, 10)


App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  
0        0      8.70        0  Paid  $13.99           Teen           Social  
1        0     33.00        0  Paid   $2.99       Everyone        Education  
2        0      5.50        0  Paid   $1.49       Everyone  Personalization  
3        0     19.00        0   NaN       0   Everyone 10+         Strategy  
4        0     10.00        0  Free       0       Everyone         Business

Drop the NaN values with .dropna()

In [ ]:
df_apps_clean = df_apps.dropna()
df_apps_clean.shape

(9367, 10)

### Find and Remove Duplicates

**Challenge**: Are there any duplicates in data? Check for duplicates using the [.duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function. How many entries can you find for the "Instagram" app? Use [.drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) to remove any duplicates from `df_apps_clean`. 


In [ ]:
duplicated_rows = df_apps_clean[df_apps_clean.duplicated()]

In [ ]:
print(duplicated_rows.shape)

(476, 10)


In [ ]:
duplicated_rows.head()

App Category  Rating  Reviews  \
946                       420 BZ Budeze Delivery  MEDICAL    5.00        2   
1133                                 MouseMingle   DATING    2.70        3   
1196  Cardiac diagnosis (heart rate, arrhythmia)  MEDICAL    4.40        8   
1231                                Sway Medical  MEDICAL    5.00        3   
1247              Chat Kids - Chat Room For Kids   DATING    4.70        6   

      Size_MBs Installs  Type   Price Content_Rating   Genres  
946      11.00      100  Free       0     Mature 17+  Medical  
1133      3.90      100  Free       0     Mature 17+   Dating  
1196      6.50      100  Paid  $12.99       Everyone  Medical  
1231     22.00      100  Free       0       Everyone  Medical  
1247      4.90      100  Free       0     Mature 17+   Dating

In [ ]:
df_apps_clean[df_apps_clean.App == 'Instagram']

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   
10808  Instagram   SOCIAL    4.50  66577446      5.30  1,000,000,000  Free   
10809  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   
10810  Instagram   SOCIAL    4.50  66509917      5.30  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social  
10808     0           Teen  Social  
10809     0           Teen  Social  
10810     0           Teen  Social

`.drop_duplicates() `works if only the entries are explicitly unique, without other columns/fields being different between the copies. 

In [ ]:
df_apps_clean = df_apps_clean.drop_duplicates()

Must specify how to identify duplicates when other column names are used in comparison to identifying the dupicates. 





In [ ]:
# need to specify the subset for identifying duplicates
df_apps_clean = df_apps_clean.drop_duplicates(subset=["App", "Type", "Price"])

# now verify a single unique record has been retained
df_apps_clean[df_apps_clean.App == 'Instagram']

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social

In [ ]:
df_apps_clean.shape

(8199, 10)

# Find Highest Rated Apps

**Challenge**: Identify which apps are the highest rated. What problem might you encounter if you rely exclusively on ratings alone to determine the quality of an app?

In [ ]:
df_apps_clean.sort_values('Rating', ascending=False).head()

App     Category  Rating  Reviews  Size_MBs Installs  \
21    KBA-EZ Health Guide      MEDICAL    5.00        4     25.00        1   
1230         Sway Medical      MEDICAL    5.00        3     22.00      100   
1227    AJ Men's Grooming    LIFESTYLE    5.00        2     22.00      100   
1224       FK Dedinje BGD       SPORTS    5.00       36      2.60      100   
1223      CB VIDEO VISION  PHOTOGRAPHY    5.00       13      2.60      100   

      Type Price Content_Rating       Genres  
21    Free     0       Everyone      Medical  
1230  Free     0       Everyone      Medical  
1227  Free     0       Everyone    Lifestyle  
1224  Free     0       Everyone       Sports  
1223  Free     0       Everyone  Photography

Only apps with very few reviews and a low number of installs have perfect 5 star ratings (most likely by friends and family).

# Find 5 Largest Apps in terms of Size (MBs)

**Challenge**: What's the size in megabytes (MB) of the largest Android apps in the Google Play Store. Based on the data, do you think there could be limit in place or can developers make apps as large as they please? 

In [ ]:
df_apps_clean.sort_values('Size_MBs', ascending=False).head()

App            Category  Rating  Reviews  \
9942   Talking Babsy Baby: Baby Games           LIFESTYLE    4.00   140995   
10687          Hungry Shark Evolution                GAME    4.50  6074334   
9943            Miami crime simulator                GAME    4.00   254518   
9944     Gangster Town: Vice District              FAMILY    4.30    65146   
3144                       Vi Trainer  HEALTH_AND_FITNESS    3.60      124   

       Size_MBs     Installs  Type Price Content_Rating  \
9942     100.00   10,000,000  Free     0       Everyone   
10687    100.00  100,000,000  Free     0           Teen   
9943     100.00   10,000,000  Free     0     Mature 17+   
9944     100.00   10,000,000  Free     0     Mature 17+   
3144     100.00        5,000  Free     0       Everyone   

                       Genres  
9942   Lifestyle;Pretend Play  
10687                  Arcade  
9943                   Action  
9944               Simulation  
3144         Health & Fitness

# Find the 5 App with Most Reviews

**Challenge**: Which apps have the highest number of reviews? Are there any paid apps among the top 50?

In [ ]:
df_apps_clean.sort_values('Reviews', ascending=False).head(5)

App       Category  Rating  \
10805                                  Facebook         SOCIAL    4.10   
10785                        WhatsApp Messenger  COMMUNICATION    4.40   
10806                                 Instagram         SOCIAL    4.50   
10784  Messenger – Text and Video Chat for Free  COMMUNICATION    4.00   
10650                            Clash of Clans           GAME    4.60   

        Reviews  Size_MBs       Installs  Type Price Content_Rating  \
10805  78158306      5.30  1,000,000,000  Free     0           Teen   
10785  69119316      3.50  1,000,000,000  Free     0       Everyone   
10806  66577313      5.30  1,000,000,000  Free     0           Teen   
10784  56642847      3.50  1,000,000,000  Free     0       Everyone   
10650  44891723     98.00    100,000,000  Free     0   Everyone 10+   

              Genres  
10805         Social  
10785  Communication  
10806         Social  
10784  Communication  
10650       Strategy

The list of the top 50 apps most reviewed on the Android App Store are FREE.

# Plotly Pie and Donut Charts - Visualise Categorical Data: Content Ratings

All Android Apps have a content rating like "Everyone" or "Teen" or "Mature 17+". 

Take a look at the distribution of the content ratings in the dataset and see how to visualise it with plotly - a popular data visualisation library and great in combination or alternative to Matplotlib.

First, count the number of occurrences of each rating with `.value_counts()`.

In [107]:
ratings = df_apps_clean.Content_Rating.value_counts()
ratings

Everyone           6607
Teen                911
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Unrated               1
Name: Content_Rating, dtype: int64

First step to creating charts with plotly is to `import` plotly.express. This is the fastest way to create a beautiful graphic with a minimal amount of code in plotly. 

In [108]:
import pandas as pd
import plotly.express as px

To create a pie chart, simply call `px.pie()` and then `.show()` the resulting figure.

Configure other aspects of the pie chart by calling the `.update_traces()` method. 

In plotly, "traces" refers to graphical marks on a figure, like a collection of attributes.

In [111]:
fig = px.pie(labels=ratings.index, 
             values=ratings.values,
             title='Content Rating',
             names=ratings.index)
fig.update_traces(textposition='outside', textfont_size=15, textinfo='percent+label')
fig.show()

To creat a donut 🍩 chart, simply add a value for the `'hole'` argument.

In [109]:
fig = px.pie(labels=ratings.index, 
             values=ratings.values,
             title='Content Rating',
             names=ratings.index,
             hole=0.6)

fig.update_traces(textposition='inside', textfont_size=15, textinfo='percent')
fig.show()

Customise the pie chart. https://plotly.com/python-api-reference/generated/plotly.express.pie.html

# Numeric Type Conversion: Examine the Number of Installs

**Challenge**: How many apps had over 1 billion (that's right - BILLION) installations? How many apps just had a single install? 

Check the datatype of the Installs column.

Count the number of apps at each level of installations. 

Convert the number of installations (the Installs column) to a numeric data type. Hint: this is a 2-step process. You'll have make sure you remove non-numeric characters first. 

In [ ]:
df_apps_clean.Installs.describe()

count          8199
unique           19
top       1,000,000
freq           1417
Name: Installs, dtype: object

To check the data types, use `.describe() `on the column (shown above)  or `.info()` on the DataFrame (below)

In [ ]:
df_apps_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8199 entries, 21 to 10835
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             8199 non-null   object 
 1   Category        8199 non-null   object 
 2   Rating          8199 non-null   float64
 3   Reviews         8199 non-null   int64  
 4   Size_MBs        8199 non-null   float64
 5   Installs        8199 non-null   object 
 6   Type            8199 non-null   object 
 7   Price           8199 non-null   object 
 8   Content_Rating  8199 non-null   object 
 9   Genres          8199 non-null   object 
dtypes: float64(2), int64(1), object(7)
memory usage: 704.6+ KB


Take 2 columns, say Installs and the App name, and count the number of entries per level of installations with .group() and .count()

Notice, we are dealing with non-numeric data type, therefore the ordering is not helpful. The reason that Python does not recognising the installs as numbers is because of the comma (,) characters. 

In [ ]:
df_apps_clean[['App', 'Installs']].groupby('Installs').count()

App
Installs           
1                 3
1,000           698
1,000,000      1417
1,000,000,000    20
10               69
10,000          988
10,000,000      933
100             303
100,000        1096
100,000,000     189
5                 9
5,000           425
5,000,000       607
50               56
50,000          457
50,000,000      202
500             199
500,000         504
500,000,000      24

Remove the comma(,) character or any character for that matter from the DataFrame using the string's **.replace()** method. This replaces every comma (,) character with an empty string. 

This completely removes all the commans in the Installs column. Then we can convert our data to a number using **.to_numeric()**

In [ ]:
df_apps_clean.Installs = df_apps_clean.Installs.astype(str).str.replace(',', "")
df_apps_clean.Installs = pd.to_numeric(df_apps_clean.Installs)
df_apps_clean[['App', 'Installs']].groupby('Installs').count()


App
Installs        
1              3
5              9
10            69
50            56
100          303
500          199
1000         698
5000         425
10000        988
50000        457
100000      1096
500000       504
1000000     1417
5000000      607
10000000     933
50000000     202
100000000    189
500000000     24
1000000000    20

# Find the Most Expensive Apps, Filter out the Junk, and Calculate a (ballpark) Sales Revenue Estimate

Let's examine the Price column more closely.

**Challenge**: Convert the price column to numeric data. Then investigate the top 20 most expensive apps in the dataset.

Remove all apps that cost more than $250 from the `df_apps_clean` DataFrame.

Add a column called 'Revenue_Estimate' to the DataFrame. This column should hold the price of the app times the number of installs. What are the top 10 highest grossing paid apps according to this estimate? Out of the top 10 highest grossing paid apps, how many are games?


In [ ]:
df_apps_clean.Price = df_apps_clean.Price.astype(str).str.replace('$', "")
df_apps_clean.Price = pd.to_numeric(df_apps_clean.Price)
df_apps_clean.sort_values('Price', ascending=False).head(20)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning:

The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.



App   Category  Rating  Reviews  Size_MBs  \
3946        I'm Rich - Trump Edition  LIFESTYLE    3.60      275      7.30   
2461              I AM RICH PRO PLUS    FINANCE    4.00       36     41.00   
4606               I Am Rich Premium    FINANCE    4.10     1867      4.70   
3145              I am rich(premium)    FINANCE    3.50      472      0.94   
3554                      💎 I'm rich  LIFESTYLE    3.80      718     26.00   
5765                       I am rich  LIFESTYLE    3.80     3547      1.80   
1946  I am rich (Most expensive app)    FINANCE    4.10      129      2.70   
2775                   I Am Rich Pro     FAMILY    4.40      201      2.70   
3221                  I am Rich Plus     FAMILY    4.00      856      8.70   
3114                       I am Rich    FINANCE    4.30      180      3.80   
1331          most expensive app (H)     FAMILY    4.30        6      1.50   
2394                      I am Rich!    FINANCE    3.80       93     22.00   
3897                       I Am Rich     FAMILY    3.60      217      4.90   
2193             I am extremely Rich  LIFESTYLE    2.90       41      2.90   
3856                   I am rich VIP  LIFESTYLE    3.80      411      2.60   
2281       Vargo Anesthesia Mega App    MEDICAL    4.60       92     32.00   
1407                    LTC AS Legal    MEDICAL    4.00        6      1.30   
2629                I am Rich Person  LIFESTYLE    4.20      134      1.80   
2481         A Manual of Acupuncture    MEDICAL    3.50      214     68.00   
4264         Golfshot Plus: Golf GPS     SPORTS    4.10     3387     25.00   

      Installs  Type  Price Content_Rating         Genres  
3946     10000  Paid 400.00       Everyone      Lifestyle  
2461      1000  Paid 399.99       Everyone        Finance  
4606     50000  Paid 399.99       Everyone        Finance  
3145      5000  Paid 399.99       Everyone        Finance  
3554     10000  Paid 399.99       Everyone      Lifestyle  
5765    100000  Paid 399.99       Everyone      Lifestyle  
1946      1000  Paid 399.99           Teen        Finance  
2775      5000  Paid 399.99       Everyone  Entertainment  
3221     10000  Paid 399.99       Everyone  Entertainment  
3114      5000  Paid 399.99       Everyone        Finance  
1331       100  Paid 399.99       Everyone  Entertainment  
2394      1000  Paid 399.99       Everyone        Finance  
3897     10000  Paid 389.99       Everyone  Entertainment  
2193      1000  Paid 379.99       Everyone      Lifestyle  
3856     10000  Paid 299.99       Everyone      Lifestyle  
2281      1000  Paid  79.99       Everyone        Medical  
1407       100  Paid  39.99       Everyone        Medical  
2629      1000  Paid  37.99       Everyone      Lifestyle  
2481      1000  Paid  33.99       Everyone        Medical  
4264     50000  Paid  29.99       Everyone         Sports

Remove bad data from our dataset so that our analysis is not misrepresented when we calculate "most expensive 'real' apps.

The top 5 apps now you can see 4 out of 5 of these are medical apps.

### The most expensive apps sub $250

In [ ]:
df_apps_clean = df_apps_clean[df_apps_clean['Price'] < 250]
df_apps_clean.sort_values('Price', ascending=False).head(5)

App   Category  Rating  Reviews  Size_MBs  \
2281  Vargo Anesthesia Mega App    MEDICAL    4.60       92     32.00   
1407               LTC AS Legal    MEDICAL    4.00        6      1.30   
2629           I am Rich Person  LIFESTYLE    4.20      134      1.80   
2481    A Manual of Acupuncture    MEDICAL    3.50      214     68.00   
2463         PTA Content Master    MEDICAL    4.20       64     41.00   

      Installs  Type  Price Content_Rating     Genres  
2281      1000  Paid  79.99       Everyone    Medical  
1407       100  Paid  39.99       Everyone    Medical  
2629      1000  Paid  37.99       Everyone  Lifestyle  
2481      1000  Paid  33.99       Everyone    Medical  
2463      1000  Paid  29.99       Everyone    Medical

### Highest Grossing Paid Apps (ballpark estimate)

Multiply the values in the price and installs column to get the vaue for highest grossing paid apps. 

Minecraft is the highest, at $70 Million in the category of Family versus Game. 

In [ ]:
df_apps_clean['Revenue_Estimate'] = df_apps_clean.Installs.mul(df_apps_clean.Price)
df_apps_clean.sort_values('Revenue_Estimate', ascending=False)[:10]

App     Category  Rating  Reviews  Size_MBs  \
9220                      Minecraft       FAMILY    4.50  2376564     19.00   
8825                  Hitman Sniper         GAME    4.60   408292     29.00   
7151  Grand Theft Auto: San Andreas         GAME    4.40   348962     26.00   
7477            Facetune - For Free  PHOTOGRAPHY    4.40    49553     48.00   
7977        Sleep as Android Unlock    LIFESTYLE    4.50    23966      0.85   
6594            DraStic DS Emulator         GAME    4.60    87766     12.00   
6082                   Weather Live      WEATHER    4.50    76593      4.75   
7954                    Bloons TD 5       FAMILY    4.60   190086     94.00   
7633        Five Nights at Freddy's         GAME    4.60   100805     50.00   
6746     Card Wars - Adventure Time       FAMILY    4.30   129603     23.00   

      Installs  Type  Price Content_Rating                     Genres  \
9220  10000000  Paid   6.99   Everyone 10+  Arcade;Action & Adventure   
8825  10000000  Paid   0.99     Mature 17+                     Action   
7151   1000000  Paid   6.99     Mature 17+                     Action   
7477   1000000  Paid   5.99       Everyone                Photography   
7977   1000000  Paid   5.99       Everyone                  Lifestyle   
6594   1000000  Paid   4.99       Everyone                     Action   
6082    500000  Paid   5.99       Everyone                    Weather   
7954   1000000  Paid   2.99       Everyone                   Strategy   
7633   1000000  Paid   2.99           Teen                     Action   
6746   1000000  Paid   2.99   Everyone 10+    Card;Action & Adventure   

      Revenue_Estimate  
9220     69,900,000.00  
8825      9,900,000.00  
7151      6,990,000.00  
7477      5,990,000.00  
7977      5,990,000.00  
6594      4,990,000.00  
6082      2,995,000.00  
7954      2,990,000.00  
7633      2,990,000.00  
6746      2,990,000.00

# Plotly Bar Charts & Scatter Plots: Analysing App Categories

Let's analyze the most competitive categories dominating the market. Find the number of different categories. 

In [ ]:
df_apps_clean.Category.nunique()

33

In [ ]:
top10_category = df_apps_clean.Category.value_counts()[:10]
top10_category

FAMILY             1606
GAME                910
TOOLS               719
PRODUCTIVITY        301
PERSONALIZATION     298
LIFESTYLE           297
FINANCE             296
MEDICAL             292
PHOTOGRAPHY         263
BUSINESS            262
Name: Category, dtype: int64

### Vertical Bar Chart - Highest Competition (Number of Apps)

Based on the number of apps, the Family and Game categories are the most competitive. Releasing yet another app into these categories will make it hard to get noticed.

In [ ]:
bar = px.bar(x = top10_category.index, y = top10_category.values)
bar.show()

### Horizontal Bar Chart - Most Popular Categories (Highest Downloads)

Look at this from a different perspective. What matters is not just the total number of apps in the category but how often apps are downloaded in that category. This will give us an idea of how popular a category is. First, group all of the apps by category and sum the number of installations. 

In [ ]:
category_installs = df_apps_clean.groupby('Category').agg({'Installs': pd.Series.sum})
category_installs.sort_values('Installs', ascending=True, inplace=True)

In [ ]:
h_bar = px.bar(x = category_installs.Installs, 
               y = category_installs.index, 
               orientation = 'h', 
               title = 'Category Popularity')

h_bar.update_layout(xaxis_title='Number of Downloads', yaxis_title='Category')
h_bar.show()

Now you can see that Games and Tools are actually the most popular categories. If we plot the popularity of a category next to the number of apps in that category we can get an idea of how concentrated a category is. Do few apps have the most downloads or are the downloads spread out over many apps? 

### Category Concentration - Downloads vs. Competition

**Challenge**: 
* First, create a DataFrame that has the number of apps in one column and the number of installs in another:
<!-- 
<img src=https://imgur.com/uQRSlXi.png width="350"> -->

* Then use the [plotly express examples from the documentation](https://plotly.com/python/line-and-scatter/) alongside the [.scatter() API reference](https://plotly.com/python-api-reference/generated/plotly.express.scatter.html)to create scatter plot that looks like this. 

<!-- <img src=https://imgur.com/cHsqh6a.png> -->

*Hint*: Use the size, hover_name and color parameters in .scatter(). To scale the yaxis, call .update_layout() and specify that the yaxis should be on a log-scale like so: yaxis=dict(type='log') 

First, work out the number of apps in each category then use .merge() and combine the two DataFrames. Then create the chart. Note: YOu have to pass an entire DataFrame and specify which columns should be used for the x and y by column name. 

In [ ]:
cat_number = df_apps_clean.groupby('Category').agg({'App': pd.Series.count})
cat_merged_df = pd.merge(cat_number, category_installs, on='Category', how="inner")
print(f'The dimensions of the DataFrame are: {cat_merged_df.shape}')
cat_merged_df.sort_values("Installs", ascending=False)

The dimensions of the DataFrame are: (33, 2)


App     Installs
Category                              
GAME                  910  13858762717
COMMUNICATION         257  11039241530
TOOLS                 719   8099724500
PRODUCTIVITY          301   5788070180
SOCIAL                203   5487841475
PHOTOGRAPHY           263   4649143130
FAMILY               1606   4437554490
VIDEO_PLAYERS         148   3916897200
TRAVEL_AND_LOCAL      187   2894859300
NEWS_AND_MAGAZINES    204   2369110650
ENTERTAINMENT         102   2113660000
BOOKS_AND_REFERENCE   169   1665791655
PERSONALIZATION       298   1532352930
SHOPPING              180   1400331540
HEALTH_AND_FITNESS    243   1134006220
SPORTS                260   1096431465
BUSINESS              262    692018120
LIFESTYLE             297    503611120
MAPS_AND_NAVIGATION   118    503267560
FINANCE               296    455249400
WEATHER                72    361096500
EDUCATION             118    352852000
FOOD_AND_DRINK         94    211677750
DATING                134    140912410
ART_AND_DESIGN         61    114233100
HOUSE_AND_HOME         62     97082000
AUTO_AND_VEHICLES      73     53129800
LIBRARIES_AND_DEMO     64     52083000
COMICS                 54     44931100
MEDICAL               292     39162676
PARENTING              50     31116110
BEAUTY                 42     26916200
EVENTS                 45     15949410

Now you can pass an entire DataFrame and specify which columns should be used for the x and y by column name.

In [ ]:
scatter = px.scatter(cat_merged_df, # data
                     x='App', # column name
                     y='Installs', 
                     title='Category Concentration',
                     size='App', 
                     hover_name=cat_merged_df.index,
                     color='Installs')
scatter.update_layout(xaxis_title="Number of Apps (Lower=More Concentrated)",
                      yaxis_title="Installs",
                      yaxis=dict(type='log'))
scatter.show()

Now you can see that categories like Family, Tools and Game have many different apps sharing a high number of downloads. But for the categories like video players and entertainment, all the downloads are concentrated in very few apps. 

# Extracting Nested Data from a Column

**Challenge**: How many different types of genres are there? Can an app belong to more than one genre? Check what happens when you use .value_counts() on a column with nested values? See if you can work around this problem by using the .split() function and the DataFrame's [.stack() method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.stack.html). 


NOTE: If you have nested data, then the len() method to get the number of unique values will not be entirely accurate. 

In [ ]:
# Get number of Genres
len(df_apps_clean.Genres.unique())

114

Using the **.value_counts()** and looking at the values that just have a a single entry. The semi-colon(;) separates the genres names. 

In [ ]:
# Problem: Have multiple categories separated by ;
df_apps_clean.Genres.value_counts().sort_values(ascending=True)[:5]

Lifestyle;Pretend Play      1
Strategy;Education          1
Adventure;Education         1
Role Playing;Brain Games    1
Tools;Education             1
Name: Genres, dtype: int64

Use the .split() method to separated teh genre names for a  more accurate calculation. 

In [ ]:
# split the strings on the semi-colon and then .stack them
stack = df_apps_clean.Genres.str.split(';', expand=True).stack()
print(f'We now have a single column with shape: {stack.shape}')
num_genres = stack.value_counts()
print(f'Number of genres: {len(num_genres)}')

We now have a single column with shape: (8564,)
Number of genres: 53


# Colour Scales in Plotly Charts - Competition in Genres

**Challenge**: Can you create the chart with the Series containing the genre data? 

<!-- <img src=https://imgur.com/DbcoQli.png width=400> -->

Try experimenting with the built in colour scales in Plotly. You can find a full list [here](https://plotly.com/python/builtin-colorscales/). 

* Find a way to set the colour scale using the color_continuous_scale parameter. 
* Find a way to make the color axis disappear by using coloraxis_showscale. 

In [ ]:
bar = px.bar(x = num_genres.index[:15], # index = category name
             y = num_genres.values[:15], # count
             title='Top Genres',
             hover_name=num_genres.index[:15], 
             color=num_genres.values[:15], 
             color_continuous_scale='Agsunset')
bar.update_layout(xaxis_title='Genre',
                  yaxis_title='Number of Apps', 
                  coloraxis_showscale=False)
bar.show()

# Grouped Bar Charts: Free vs. Paid Apps per Category

What is the split between free and paid apps. Using the total number of apps per category and the total number of apps per genre data. 

In [ ]:
df_apps_clean.Type.value_counts()

Free    7595
Paid     589
Name: Type, dtype: int64

The majority of apps are FREE on the Google Play Store. Maybe there's categories that have more paid apps than others? 

Group the data by Category and then by type. Then add up the number of apps per each type. 

**as_index=False will push the data into columns rather than end up with Categories as the index. **

In [ ]:
df_free_vs_paid = df_apps_clean.groupby(["Category", "Type"], as_index=False).agg({'App': pd.Series.count})
df_free_vs_paid.head()
# df_free_vs_paid.tail()

Category  Type  App
0     ART_AND_DESIGN  Free   58
1     ART_AND_DESIGN  Paid    3
2  AUTO_AND_VEHICLES  Free   72
3  AUTO_AND_VEHICLES  Paid    1
4             BEAUTY  Free   42

**Challenge**: Use the plotly express bar [chart examples](https://plotly.com/python/bar-charts/#bar-chart-with-sorted-or-ordered-categories) and the [.bar() API reference](https://plotly.com/python-api-reference/generated/plotly.express.bar.html#plotly.express.bar) to create the bar chart: 

<!-- <img src=https://imgur.com/LE0XCxA.png> -->

You'll want to use the `df_free_vs_paid` DataFrame that you created above that has the total number of free and paid apps per category. 

See if you can figure out how to get the look above by changing the `categoryorder` to 'total descending' as outlined in the documentation here [here](https://plotly.com/python/categorical-axes/#automatically-sorting-categories-by-name-or-total-value). 

The key is to use the color and barmode parameters for the .bar() method. To get a particular order, pass a dictionary to teh axis parameter in .update_layout()

In [ ]:
g_bar = px.bar(df_free_vs_paid, 
               x='Category', 
               y='App',
               title='Free vs Paid Apps by Category',
               color='Type',
               barmode='group')
g_bar.update_layout(xaxis_title='Category',
                    yaxis_title='Number of Apps',
                    xaxis={'categoryorder': 'total descending'},
                    yaxis=dict(type='log'))
g_bar.show()

There are very few apps on the Google Play Store, some categories have relatively more paid apps than others, including Personalization, Medical, and Weather. Depending on the category you are targeting, it might make sense to release a paid-for app. 

But this makes you wonder ...: 

1.   How much should you charge? What are other apps charging in that category?
2.   how much revenue could you make? How many downloads are you potentially giving up when your app is not free?

Analysis with Box Plots ....



# Plotly Box Plots: Lost Downloads for Paid Apps

**Challenge**: Create a box plot that shows the number of Installs for free versus paid apps. How does the median number of installations compare? Is the difference large or small?

Use the [Box Plots Guide](https://plotly.com/python/box-plots/) and the [.box API reference](https://plotly.com/python-api-reference/generated/plotly.express.box.html) to create the box plot. 
<!-- 
<img src=https://imgur.com/uVsECT3.png> -->


In [ ]:
box = px.box(df_apps_clean,
             y='Installs',
             x='Type',
             color='Type',
             notched=True,
             points='all',
             title="How Many Downloads are Paid Apps Giving Up?")
box.update_layout(yaxis=dict(type='log'))
box.show()

# Plotly Box Plots: Revenue by App Category

**Challenge**: See if you can generate a next box plot.

<!-- <img src=https://imgur.com/v4CiNqX.png> -->

Looking at the hover text, how much does the median app earn in the Tools category? If developing an Android app costs $30,000 or thereabouts, does the average photography app recoup its development costs?

Hint: Use 'min ascending' to sort the categories. 

In [ ]:
df_paid_apps = df_apps_clean[df_apps_clean['Type'] == 'Paid']
box = px.box(df_paid_apps,
             x='Category',
             y='Revenue_Estimate',
             title='How Much Can Paid Apps Earn?')
box.update_layout(xaxis_title='Category',
                  xaxis={'categoryorder':'min ascending'},
                  yaxis=dict(type='log'))
box.show()

An Android app costs 30K to develop, then the ave app in very few categorie would cover that development cost. The median photography app earned 20K. Many more app's revenues were lower - meaning they would need other sources of revenue like advertising or in-app purchases to make up for their development costs. 

Notice the high earning outliers: certain app categories that have much higher estimated revenues such as Medical, Personalisation, Tools, Game, and Family. 

# How Much Can You Charge? Examine Paid App Pricing Strategies by Category

**Challenge**: What is the median price price for a paid app? Then compare pricing by category by creating another box plot. But this time examine the prices (instead of the revenue estimates) of the paid apps. I recommend using `{categoryorder':'max descending'}` to sort the categories.

In [ ]:
df_paid_apps.Price.median()

2.99

In [ ]:
box = px.box(df_paid_apps,
             x='Category',
             y='Price',
             title='Price Per Category')

box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Price',
                  xaxis={'categoryorder': 'max descending'},
                  yaxis=dict(type='log'))
box.show()

Some categories have higher median prices than others. Medical apps have the most expensive apps as well as a median price of ﹩5.49. 

In contrast, Personalisation apps are quite cheap on average at ﹩1.49, whereas Business (﹩4.99) and Dating (﹩6.99) categories have higher median prices. Customers who shop in these categories might not be so concerend about paying a bit extra for their apps. 